### PR TITLE
Switch to leftpadding self.shape in coo-1d arrays when using 2d tools

### PR DESCRIPTION
**Pull request by dschult**
This impacts `toarray` and `add_dense` (treatments for coo_todense), `row` and `col` properties.
It essentially makes a 1d-array like a row array, instead of like a column array. This is how numpy handles 1d in the sense that it left-pads `shape` with 1s for new dimensions in broadcasting. This will simplify matching with dok, csc, csr 1d-arrays.

Also, in english (no affect on code), the 1d coo_array indices can be though of as columns. We build more dimensions by pre-pending the new index array to the `indices` tuple.  In code, handling conversion of `self.shape` to `M, N` (when using sparsetools that intake M, N) with:  `M, N = (self.shape[0] if self.ndim > 1 else 1), self.shape[-1]`

These aren't needed to pass any tests per se, but make matching to the dok, csc, csr needs easier. They need this left-pad rule for using the existing sparsetools for indexing/mul/matmul/etc.

### CODE REVIEW

File: scipy/sparse/_coo.py [hunk 1]
```diff
@@ -304,7 +305,7 @@ def tocsc(self, copy=False):
         else:
             M,N = self.shape
             idx_dtype = self._get_index_dtype(
-                (self.col, self.row), maxval=max(self.nnz, M)
+                (self.col, self.row), maxval=max(self.nnz, *self.shape)
```
Comment (perimosocordiae) [inline scipy/sparse/_coo.py]: Why was this change needed?

File: scipy/sparse/_coo.py [hunk 2]
```diff
@@ -528,8 +529,8 @@ def _add_dense(self, other):
         dtype = upcast_char(self.dtype.char, other.dtype.char)
         result = np.array(other, dtype=dtype, copy=True)
         fortran = int(result.flags.f_contiguous)
-        M = self.shape[0]
-        N = self.shape[1] if self.ndim > 1 else 1
+        M = self.shape[0] if self.ndim > 1 else 1
+        N = self.shape[-1]
```
Comment (perimosocordiae) [inline scipy/sparse/_coo.py]: Alternately:

```
*_, M, N = (1, 1) + self.shape
```

Maybe we should have private helper method for getting `(M, N)`, so we don't have to remember the trick every time.

File: scipy/sparse/_coo.py [hunk 3]
```diff
@@ -99,23 +99,26 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
 
     @property
     def row(self):
-        return self.indices[0]
+        if self.ndim > 1:
+            return self.indices[0]
+        return np.zeros(self.nnz, dtype=self.indices[0].dtype)
```
Comment (perimosocordiae) [inline scipy/sparse/_coo.py]: I'd prefer:

```
return self.indices[-2] if self.ndim > 1 else np.zeros_like(self.col)
```

File: scipy/sparse/_coo.py [hunk 4]
```diff
@@ -99,23 +99,26 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
...
+
 
     @row.setter
     def row(self, new_row):
+        if self.ndim < 2:
+            raise ValueError('cannot set row attribute of a 1-dimensional sparse array')
         new_row = np.asarray(new_row, dtype=self.indices[0].dtype)
         self.indices = (new_row,) + self.indices[1:]
```
Comment (perimosocordiae) [inline scipy/sparse/_coo.py]: I think there's a subtlety here that we need to get right: in this new system, "col" is always `indices[-1]`, "row" is `indices[-2]`, and any theoretical other dimensions keep going right to left.

So instead of using `indices[0]` for the row in the 2d case, I'd prefer if we consistently used negative indices. Then the final assignment here would become:

```
self.indices = self.indices[:-2] + (new_row, self.indices[-1])
```

File: scipy/sparse/tests/test_coo.py [hunk 5]
```diff
@@ -229,15 +229,16 @@ def test_eliminate_zeros():
...
+    assert np.array_equal(arr1d.row, np.array([0]))
 
 
 def test_1d_add_dense():
     den_a = np.array([0, -2, -3, 0])
     den_b = np.array([0, 1, 2, 3])
     exp = den_a + den_b
     res = coo_array(den_a) + den_b
-    assert type(res) == type(exp)
+    assert isinstance(res, np.ndarray)
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_coo.py]: Why change this?

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
Sorry this took me a while to get to. Let me know what you think about my negative indexing idea.

File: scipy/sparse/tests/test_coo.py (hunk 5 above)
Comment (dschult) [reply scipy/sparse/tests/test_coo.py]: I can't always keep in my head what type the numpy version creates. Testing equality with the dense version doesn't say what the result should be -- only that `res` matches `exp`. As I look at this again, I guess the `isinstance` version would pass even if one was a subclass of the other type. But I hadn't considered that before now.

I'm fine with going back to the original, but would like there to be some way to indicate what the type should be -- not just that they are equal. Perhaps: `type(res) == type(exp) == np.ndarray`?  But that feels like it is testing numpy (the `exp` part of the test).  And I'm fine with whatever we do here. I can change it back to the original and work to learn the numpy result types better. :)

File: scipy/sparse/_coo.py (hunk 4 above)
Comment (dschult) [reply scipy/sparse/_coo.py]: Good -- I agree that this is the way to nd in the long term. Adding ndim to each lookup (to handle the negative index) shouldn't be costly. And this would provide uniformity for handling these shape-related features of the arrays.

File: scipy/sparse/_coo.py (hunk 3 above)
Comment (dschult) [reply scipy/sparse/_coo.py]: Good...let's use index `-2` throughout and inline if/else in cases like this.

File: scipy/sparse/_coo.py (hunk 2 above)
Comment (dschult) [reply scipy/sparse/_coo.py]: I also thought about have a helper function, but there do seem to be multiple desires:
- finding out if this object is 2d or 1d
- reporting separately the length of each dimension
- reporting the 2d-version of the length of each dimension (for calling the sparsetools functions)

Perhaps we should have an attribute/property `self._shape2d`.  This would/will be needed with nd representations too -- at least the ones that represent the nd-sparse as a 2d-sparse like GCRS/GCRC or others. And it would do much of what these idioms here would do.

So, we could use have 3 properties/attributes `ndim` to tell 1d from 2d, `shape` to give the actual shape, and `shape2d` to give the 2d representation needed for  sparsetools functions.

File: scipy/sparse/_coo.py (hunk 1 above)
Comment (dschult) [reply scipy/sparse/_coo.py]: If self.shape is e.g. `(5, 2**33)` and no nonzero values have index higher than `(2**32 -1)` then the current code will make `idx_type` 32-bit while it needs to be 64-bit. right?  
The current code takes into account large number of rows, but not large number of columns.

I think this is a bug in the 2d code. I could make it a separate PR if that is better. And I guess it should have a test which I haven't put into this either.  Would this be better as a separate PR?

**Comment by dschult** [comment]
Thanks for the review. I've made 3 new changes. 
- Changed M -> A in a place where M is used as a matrix. This could be confusing with M being # of rows. A is for array
- updated the row property to use negative index
- added a unified way to get a 2d version of M, N to use with `sparsetools`.
   This way is a private attribute `_shape2d`. This must be set whenever `self._shape` is set directly. Perhaps we should add a property/method to update both _shape and _shape2d.
  We could instead create a method to calculate the _shape2d on the fly each time.

I'll leave the following Qs until your response:
- `assert type(res) == test(exp)`
- whether to move fixing the 2d bug for including both parts of shape in `maxval` to a separate PR

**Comment by dschult** [comment]
I like and agree with your suggestion for always using index `[-1]` for columns and `[-2]` for rows. 

For unified 2d-shape idioms, I've come to prefer a **property/method** rather than an attribute. The information is all there in `A._shape` and storing it twice will lead to being out of sync.

Name?  I've used `_shape_as_2d` here.  But `_shape2d` is another choice. Others?

In an attempt to streamline this PR, I have pulled out the following parts that are not directly related to 1d coo. 
- A->M in some code for a matrix
- assert test(res) == test(exp)
- ensuring `coo.tocsr/c` gets the right `idx_dtype`.

What should remain is:
- switch row/col logic to left pad `shape` with 1s.
- use a unified/centralized idiom for getting the 2d-equivalent shape info.

**Review by perimosocordiae — APPROVED** [review]
